zerver/worker/queue_processors.py

```diff
@@ -477,17 +477,7 @@ def consume_batch(self, user_activity_events: List[Dict[str, Any]]) -> None:
         for event in user_activity_events:
             user_profile_id = event["user_profile_id"]
 
-            if "client_id" not in event:
-                # This is for compatibility with older events still stuck in the queue,
-                # that used the client name in event["client"] instead of having
-                # event["client_id"] directly.
-                # TODO: This can be deleted for release >= 4.0.
-                if event["client"] not in self.client_id_map:
-                    client = get_client(event["client"])
-                    self.client_id_map[event["client"]] = client.id
-                client_id = self.client_id_map[event["client"]]
-            else:
-                client_id = event["client_id"]
+            client_id = event["client_id"]
```

Comment:
We should reference 200ce821a2172d82170c5357c5b32eba9d369f16.  Also, what's the reasoning around being able to remove this code in this release?

Reply:
(If users can upgrade directly without an intermediate release, we probably want to keep this)

Reply:
Hmm, we've deleted such code in the past when it was sufficiently old. How many releases do we want between a change to an event format deletion of back-compact code from the queue workers? We should update the comment on this block to accurately reflect the to delete this in - if we want to keep in 4.0.

Reply:
Our historical definition is if if there's no overlap in operating system support, so users have to chain multiple upgrades, just because that hasn't been the case for that long.

(Basically, the downside of deleting this code too early is someone directly upgrades from v1 to v2 on the same OS, and then gets weird error messages caused by the new queue handler code not supporting the events in the queue, freaks out, has a bad experience, and also generates support burden for us where we have to tell them how to clear out the unparseable events).